만국박람회 [STEP 3] 재재, 보리사랑

ZZBAE opened the pull request:
@ehgud0670   제이슨 안녕하세요 🥰
만국박람회 step2 리뷰 반영과 함께 step3 진행 완료 후 PR 보냅니다!
잘 부탁드립니다 🫡


# 만국박람회 Step 3
> 프로젝트 기간 : 2022.06.20 ~ 2022.06.22
> 팀원 : 재재, 보리사랑
> 리뷰어 : 제이슨

## 개발환경 및 라이브러리
[![swift](https://img.shields.io/badge/swift-5.6-orange)]() [![xcode](https://img.shields.io/badge/Xcode-13.3.1-blue)]()

## Step 3 수행 중 핵심경험
- [x] 오토 레이아웃을 적용하여 다양한 기기에 대응
- [x] 접근성(Accessibility)의 개념과 필요성 이해
- [x] Dynamic Types를 통해 텍스트 접근성 향상


## TroubleShooting
### 1. **Image View의 크기가 고정되지 않고 늘어나는 이슈**
최초에 Image View의 크기를 직접 정하지 않고 top, leading, bottom은 스택뷰에, trailing은 label을 담는 verticalStack View에 맞췄습니다. 

![](https://i.imgur.com/tq56Xv3.png)

다만 이렇게 설정하고 나니 imageView의 원본이 되는 asset의 크기가 매우 클 경우 Label에 굉장히 여백이 많이 들어가서 셀이 비정상적으로 출력되는 현상이 있었습니다. 

이를 해결하고자 접근을 다르게 하여 기존에는 label을 담는 Vertical Stackview의 크기를 superview의 크기에 비례하게 고정했으나 반대로 **ImageView의 크기를 고정시켜 주었고**, vertical stackview를 이에 맞춰주었습니다. ImageView의 크기를 고정으로(aspect ratio 1:1로) 지정할 경우 위의 사진처럼 해금의 사진이 정사각형으로 늘어나서 사용하면 안된다고 생각했었는데 이번 기회를 통해 **원본 비율을 유지하며 출력**된다는걸 알 수 있었습니다. 

이렇게 하니 정상적으로 출력이 되어 해결할 수 있었습니다. 

### 2. **ExpoHomeViewController의 화면을 고정시키기**

Step3 요구사항 첫번째에 첫 화면은 세로로만 볼 수 있도록 하는 조건이 있어서, 첫번째 뷰컨인 ExpoHomeViewController만 화면 회전 고정을 시켜주고 나머지 ExpoMenuViewController랑 DetailViewController는 회전을 해도 정상적으로 표시될 수 있게 해주었습니다.

우선 ExpoHomeViewController의 화면 전환 제어를 위해 AppDelegate에는 밑의 메서드를,
```swift
class AppDelegate: UIResponder, UIApplicationDelegate {
    var orientationLock = UIInterfaceOrientationMask.all

    func application(_ application: UIApplication, didDiscardSceneSessions sceneSessions: Set<UISceneSession>) {}

    func application(_ application: UIApplication, supportedInterfaceOrientationsFor window: UIWindow?) -> UIInterfaceOrientationMask {
            return self.orientationLock
    }
}

struct AppUtility {
    static func lockOrientation(_ orientation: UIInterfaceOrientationMask) {
        
        if let delegate = UIApplication.shared.delegate as? AppDelegate {
            delegate.orientationLock = orientation
        }
    }
    
    static func lockOrientation(_ orientation: UIInterfaceOrientationMask, andRotateTo rotateOrientation:UIInterfaceOrientation) {
        
        self.lockOrientation(orientation)
        
        UIDevice.current.setValue(rotateOrientation.rawValue, forKey: "orientation")
        UINavigationController.attemptRotationToDeviceOrientation()
    }
```
그리고 ExpoHomeViewController에서는 portrait상태로만 존재하고, push 된 하위 뷰 컨트롤러에서는 모두 출력되게 하기 위해 ExpoHomeViewcontroller의 viewWillAppear 함수 안에 밑의 코드를 추가해주었고
```swift
AppUtility.lockOrientation(.portrait)
```
viewWillDisappear 메서드 안에는 다음과 같이 설정하게 해주었습니다. 
```swift 
AppUtility.lockOrientation(.all)
```


## 조언을 구하고 싶은 부분
### 1. **DetailViewController의 description이 한줄로 나오는 이슈**

| 설명  | 화면 구현                                 |
| :----: |:-----------------------------------------:|
| DetailViewController의 컨텐츠는 Vertical 스택뷰 내부의 imageView와 description 라벨 두가지로 구성되어 있습니다. 이때 각 컨텐츠들이 본연의 크기대로 정렬되게 하고자 스택뷰의 distribution을 `Fill Proportionally`로 지정해 주었는데 이렇게 하니 다음과 같이 특정 컨텐츠에 대해서만 한줄로 나오게 되었습니다. </br> Distribution을 `Fill`로 변경하니 문제가 해결되었으나, `Fill Proportionally`를 적용하여 한줄만 출력이 된다면, 모든 Description 라벨이 한줄로 출력되어야 한다고 생각을 했습니다. Fill과 Fill proportionally에 이렇게 되는 차이점이 있는건지 궁금합니다. | <img src=https://i.imgur.com/MveNmDn.gif width=2500> |

### 2. **storyboard로 프로젝트 상세 구현은 성공하였지만, 코드로 작성하지 못한 아쉬움**
저희가 이번 만국박람회 프로젝트를 진행하면서, 세개의 뷰컨을 띄우는게 처음이라 코드보다는 스토리보드로 전반적인 흐름이나 구현 방식을 이해를 먼저 하는게 좋을 것 같다는 의견이 맞아 스토리보드로 진행을 하게 되었습니다.<br>
결과물은 만족스럽지만 아무래도 코드로 작성하지 못한 아쉬움이 있어, 나중에라도 이번 만국박람회 프로젝트를 코드로 작성해봤으면 좋겠다는 생각이 있습니다. 코드 작성 시 유의해야할 사항이나 제이슨의 꿀팁 등등의 의견을 듣고 싶습니다!

---
## 커밋 규칙
* feat    : 기능 추가 (새로운 기능)
* fix     : 버그 수정
* refactor: 리팩토링 (네이밍 수정 등)
* style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
* docs    : 문서 변경 (문서 추가, 수정, 삭제)
* test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)

## 코드 컨벤션
Swift Statements 형식 준수

ZZBAE committed:
docs: update README.md

- 만국박람회 프로젝트 step1 리드미 추가

ZZBAE committed:
feat: 각 ViewController 에 Dynamic Type 적용

- ExpoHomeViewController, ExpoMenuViewController, DetailViewController의 이미지와 label마다 Dynamic Type 설정
- Title label은 Title1 / Title3으로, label은 body로 지정

yusw10 committed:
refactor: ExpoHomeViewController Layout 구성

- 스크롤뷰의 top, bottom이 safe area가 아니라 superView를 따라서 전체에 나오게 변경
- contentView의 height을 고정값에서 'Greater than or equal' 으로 변경하여 컨텐츠가 모두 출력되도록 수정
- description라벨의 좌우 마진 설정
- visitors 라벨을 decimal 스타일로 변경
- 라벨 앞쪽에 어떤 라벨인지 설명하는 string 추가
- title 라벨 중간에 줄바꿈 추가

ZZBAE committed:
refactor: ExpoMenuViewController와 DetailViewController의 오토레이아웃 설정

- ExpoMenuViewController의 ExpoItemImageView의 크기 aspect ratio 1:1 로 설정
- ExpoItemNameLabel, ExpoItemDescriptionLabel 의 위치와 크기 설정
- DetailViewController의 ExpoItemImageView와 ExpoItemDescriptionLabel 크기 설정

yusw10 committed:
refactor: DetailViewController의 Layout 재설정

- DescriptionLabel의 contentHuggingpriority 250으로 설정
- 이미지뷰와 DescriptionLabel을 묶는 verticalStackview의 distribution을 fill로 변경
- DescriptionLabel의 좌우 margin추가

ZZBAE committed:
feat: ExpoHomeViewController의 화면 전환 제어 기능 구현

- AppDelegate에 lookOrientation 메서드를 생성하여 AppUtility 구조체를 사용하여 ExpoHomeViewController의 화면 전환을 제어하고,
portrait을 유지하게끔 구현

ZZBAE committed:
refactor: Step2 추가 코멘트 기반 refactoring

- IBOutlet 프로퍼티의 네이밍 수정 (소문자로 시작하는 컨벤션 준수)
- ExpoMenuTableViewCell에서 setupUI 메서드 추가

ehgud0670 commented on Expo1900/Expo1900/AppDelegate.swift:
```diff
@@ -16,5 +18,25 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     func application(_ application: UIApplication, didDiscardSceneSessions sceneSessions: Set<UISceneSession>) {}
+
+    func application(_ application: UIApplication, supportedInterfaceOrientationsFor window: UIWindow?) -> UIInterfaceOrientationMask {
+            return self.orientationLock
```
```suggestion
        return self.orientationLock
```

ehgud0670 commented on Expo1900/Expo1900/Controller/DetailViewController.swift:
```diff
@@ -21,7 +21,7 @@ final class DetailViewController: UIViewController {
-        ExpoItemImageView.image = UIImage(named: expoItemElement?.imageName ?? "")
-        ExpoItemDescriptionLabel.text = expoItemElement?.desc
+        expoItemImageView.image = UIImage(named: expoItemElement?.imageName ?? "")
+        expoItemDescriptionLabel.text = expoItemElement?.desc
```
소문자로 반영 잘 해주셨습니다 👍

ehgud0670 commented on Expo1900/Expo1900/View/ExpoMenuTableViewCell.swift:
```diff
@@ -11,6 +11,11 @@ class ExpoMenuTableViewCell: UITableViewCell {
     @IBOutlet weak var expoItemNameLabel: UILabel!
     @IBOutlet weak var expoItemDescriptionLabel: UILabel!
+   
+    func setupUI(with expoItemElement: ExpoItemElement) {
```
음, 만드셨으면 직접 써도 되지 않을까 합니다 ㅎㅎ

ehgud0670 replied:
이 메소드만 있으면 모델값을 인수로 전달받아서 뷰를 setup 하는거라 외부에서 코드 한줄만 호출하면 됩니다. 즉 코드가 깔끔해집니다.
또 내부에서만 호출되는 메소드들은 private 으로 둬도 되고요.

ZZBAE replied:
오 정말 깔끔해졌습니다! setupUI 메서드 사용해주었습니다! ㅎ ㅎ

ehgud0670 commented:
<img width="270" alt="스크린샷 2022-06-23 오후 7 39 33" src="https://user-images.githubusercontent.com/38216027/175280525-c6c3acd6-7651-4e04-b390-c5822ab2f34f.png">

여기(ExpoMenuViewController) 레이아웃 이슈 해결해주셔야 할 것 같아요~

ehgud0670 commented:
HomeVC도 navigationController 안에 같이 들어갈 필요가 있을지 궁금합니다.
ExpoMenuVC 부터 navigationController 을 사용해도 되지 않을까 해요.
그리고 HomeVC가 safeArea에 맞춰지지 않아 있어서 노치(ex. iphone12)인 폰에 콘텐츠가 넘쳐보여요(top).

<img width="400" alt="스크린샷 2022-06-23 오후 7 49 13" src="https://user-images.githubusercontent.com/38216027/175282626-8491de59-34a5-423e-abc1-7c56a1382ff2.png">

ehgud0670 commented on Expo1900/Expo1900/Controller/ExpoHomeViewController.swift:
```diff
@@ -25,21 +25,43 @@ class ExpoHomeViewController: UIViewController {
         descriptionLabel.text = exposition?.expositionDescription
         leftFlagImageView.image = UIImage(named: "flag")
         rightFlagImageView.image = UIImage(named: "flag")
+        navigationController?.navigationBar.isHidden = true
```
viewWillAppear에서 호출하고 있는데 여기도 같이 있네요. viewWillAppear에서만 히든처리해도 될 듯 합니다. 
아까도 말씀드렸지만 HomeVC은 굳이 navigationController가 적용되어 있지 않아도 될 듯 한데 어떻게 생각하시나요?
(SafeArea 적용하는데도 걸림돌이 되는 것 같아서요)

ZZBAE replied:
제이슨 말씀대로 생각해보니 정말 굳이 설정해준 것 같다고 생각이 들어서 적용해제해주었습니다. 감사합니다!

ehgud0670 commented on Expo1900/Expo1900/Controller/ExpoHomeViewController.swift:
```diff
@@ -25,21 +25,43 @@ class ExpoHomeViewController: UIViewController {
+    
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        AppUtility.lockOrientation(.all)
```
이 처리는 HomeVC가 아닌 orientation이 .all인 화면에 적용해야 하지 않을까요~?

ZZBAE replied:
넘어가는 화면 쪽에서 설정해주면 AppDelegate의 설정을 바꾸는거라 portrait만 되는 상황에서 MenuVC으로 넘어가면 설정 그대로 화면이 구성될 것 같아서 HomeVC에다가 해주었는데, orientation이 .all인 화면에 해주어도 괜찮을까요?

ehgud0670 replied:
네 그게 더 자연스러운 것 같아요. 홈에서 portrait으로 있어야 할 화면을 띄워줘야 하는 
화면이 추가로 생기면 viewWillDisappear에서의 AppUtility.lockOrientation(.all) 설정은 코드 수정해야 합니다. 
제 얘기는 HomeVC와 독립되게 설정해야 하는 코드라면 HomeVC 말고 해당 설정이 필요한 화면에서 해야 한다는 거에요 😄

ehgud0670 commented on Expo1900/Expo1900/AppDelegate.swift:
```diff
@@ -16,5 +18,25 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        }
+    }
+    
+    static func lockOrientation(_ orientation: UIInterfaceOrientationMask, andRotateTo rotateOrientation:UIInterfaceOrientation) {
```
이건 사용하지 않는 메소드 같은데 왜 같이 구현하시는지 궁금합니다 :)

ehgud0670 commented on Expo1900/Expo1900/AppDelegate.swift:
```diff
@@ -16,5 +18,25 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
+struct AppUtility {
+    static func lockOrientation(_ orientation: UIInterfaceOrientationMask) {
+        
+        if let delegate = UIApplication.shared.delegate as? AppDelegate {
+            delegate.orientationLock = orientation
+        }
```
```suggestion
        if let delegate = UIApplication.shared.delegate as? AppDelegate {
            delegate.orientationLock = orientation
        }
```

ehgud0670 reviewed (commented):
계속 리뷰 드릴게요~

ehgud0670 commented on Expo1900/Expo1900/AppDelegate.swift:
```diff
@@ -16,5 +18,25 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        }
+    }
+    
+    static func lockOrientation(_ orientation: UIInterfaceOrientationMask, andRotateTo rotateOrientation:UIInterfaceOrientation) {
```
사용안하는데 왜 만드셨는지 궁금하네요~

ZZBAE replied:
작동원리를 공부하다가 테스트로 넣어본건데 그대로 commit 해버렸네요 실수입니다 죄송합니다ㅠㅠ!! 삭제해주었습니다!

ehgud0670 replied:
앗 죄송할건 아닙니다. 그냥 궁금해서 여쭤 본겁니다 ㅠ 
테스트 용도로 넣을 수 있죠! 크게 신경쓰지 않으셔도 됩니다.

ehgud0670 commented:
### 1. DetailViewController의 description이 한줄로 나오는 이슈

=> stackView의 element 각각이 정확한 heightAnchor 값이 적용되어야 fillProportionally가 잘 적용됩니다.
저는 sizeThatFits와 heightAnchor 로 해결하긴 했어요, 참고차 말씀드려요 ㅎㅎ

### 2. storyboard로 프로젝트 상세 구현은 성공하였지만, 코드로 작성하지 못한 아쉬움

여러분들이 직접 코드로 구현해서 시행착오를 겪어가며 여러분만의 노하우를 갖는게 꿀팁입니다. 결국에는 자신이 해결하는게 가장 큰 자산이더라고요 ㅎㅎ 

피드백 보시고 의문이나 궁금한 점 있으면 말씀주세요! 고생하셨습니다~! 😁

yusw10 committed:
docs: README.md update

- Step별 상세 내용 파일 첨부

ZZBAE committed:
Update README.md

ZZBAE committed:
Merge branch 'main' into step3

ZZBAE committed:
refactor: ExpoMenuViewController의 Label 수정

- ExpoItemNameLabel의 Font 종류 변경 (Title 3 -> 2)
- ExpoItemNameLabel의 height 설정하여 오류 잡아줌

ZZBAE committed:
Merge branch 'step3' of https://github.com/ZZBAE/ios-exposition-universelle into step3

ZZBAE committed:
refactor: 리뷰 반영 컨벤션 정리

Co-authored-by: Doe-Hyung KIM <kimdo2297@gmail.com>

ZZBAE committed:
refactor: 리뷰반영 컨벤션 정리

Co-authored-by: Doe-Hyung KIM <kimdo2297@gmail.com>

ZZBAE commented:
@ehgud0670 만국박람회 요구명세서에서 홈뷰컨에 이런식으로 나와있어서 일부러 노치를 넘어가게 설정해주었습니다!

<img width="588" alt="스크린샷 2022-06-23 오후 8 15 22" src="https://user-images.githubusercontent.com/98302604/175503110-e1fb6d03-0d59-4327-ac90-de75ce74d5cb.png">

ZZBAE committed:
refactor: 리뷰 반영 AppDelegate에서 사용하지 않는 메서드 정리

- AppUtility에서 lookOrientation 삭제

ZZBAE committed:
Merge branch 'step3' of https://github.com/ZZBAE/ios-exposition-universelle into step3

ehgud0670 commented:
네, 제가 말씀 드리는 건 여러분들이 구현하신건 top 쪽이 safeArea에 맞춰진게 아니라 superview에 맞춰진게 아닌가 해서요~!
타이틀(파리_만국박람회)이 위로 올라가있네요!


<img width="534" alt="스크린샷 2022-06-24 오후 6 20 17" src="https://user-images.githubusercontent.com/38216027/175505556-5872e07f-6988-45e2-be61-99b70344c516.png">

ZZBAE committed:
refactor: ExpoMenuViewController 에서 setupUI 메서드 사용

- ExpoMenuTableViewCell에서 만들어준 setupUI 메서드를 ExpoMenuViewController에서 사용해주게끔 변경

ZZBAE committed:
refactor: ExpoHomeViewController에 NavigationBar 리뷰 반영

- ExpoHomeViewController에서 NavigationBar를 설정해준 라인 삭제 (굳이 필요 없음)

ehgud0670 reviewed (approved):
수고하셨습니다 👏
